src/server/coprocessor/executor/index_scan.rs

```diff
@@ -0,0 +1,260 @@
+               statistics: &'a mut Statistics,
+               start_ts: u64)
+               -> IndexScanExec<'a> {
+        let mut pk_col = None;
+        let desc = meta.get_desc();
+        let mut cols = meta.mut_columns();
+        if cols.last().map_or(false, |c| c.get_pk_handle()) {
+            pk_col = Some(cols.pop().unwrap());
+        }
+        let col_ids = cols.iter()
+            .map(|c| c.get_column_id())
+            .collect();
+        let scanner = Scanner::new(desc, false, snapshot, statistics, start_ts);
+        IndexScanExec {
+            desc: desc,
+            col_ids: col_ids,
+            scanner: scanner,
+            key_ranges: key_ranges,
+            cursor: Default::default(),
+            pk_col: pk_col,
+        }
+    }
+
+    pub fn get_row_from_range(&mut self) -> Result<Option<Row>> {
+        let range = &self.key_ranges[self.cursor];
+        if range.get_start() > range.get_end() {
+            return Ok(None);
+        }
+        let kv = try!(self.scanner.next_row(range));
+        let (key, value) = match kv {
+            Some((key, value)) => (key, value),
+            None => return Ok(None),
+        };
+
+        let seek_key = if self.desc {
+            key.clone()
+        } else {
+            prefix_next(&key)
+        };
+        self.scanner.set_seek_key(Some(seek_key));
+
+        let (mut values, handle) = {
+            box_try!(table::cut_idx_key(key, &self.col_ids))
+        };
+
+        let handle = if handle.is_none() {
+            box_try!(value.as_slice().read_i64::<BigEndian>())
+        } else {
+            handle.unwrap()
+        };
+
+        if let Some(ref pk_col) = self.pk_col {
+            let handle_datum = if mysql::has_unsigned_flag(pk_col.get_flag() as u64) {
+                // PK column is unsigned
+                datum::Datum::U64(handle as u64)
+            } else {
+                datum::Datum::I64(handle)
+            };
+            let mut bytes = datum::encode_value(&[handle_datum]).unwrap();
+            values.append(pk_col.get_column_id(), &mut bytes);
+        }
+        Ok(Some(Row::new(handle, values)))
+    }
+}
+
+impl<'a> Executor for IndexScanExec<'a> {
+    fn next(&mut self) -> Result<Option<Row>> {
+        while self.cursor < self.key_ranges.len() {
+            let data = box_try!(self.get_row_from_range());
+            if data.is_none() {
+                self.scanner.set_seek_key(None);
+                self.cursor += 1;
+                continue;
+            }
+            return Ok(data);
+        }
+        Ok(None)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use super::super::scanner::test::{Data, TestStore, prepare_index_data, get_idx_range};
+    use std::i64;
+    use tipb::schema::ColumnInfo;
+    use storage::Statistics;
+    use protobuf::RepeatedField;
+
+    const TABLE_ID: i64 = 1;
+    const INDEX_ID: i64 = 1;
+    const KEY_NUMBER: usize = 10;
+
+    struct IndexScanExecutorMeta {
```

Comment:
Why name it `Meta`?
Isn't it conflict with line 37?

Please choose a better name.